spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/StacktraceFingerprintingRecoveryStrategy.java

```diff
@@ -0,0 +1,136 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.completion;
+
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.springframework.cloud.dataflow.core.StreamDefinition;
+import org.springframework.util.Assert;
+
+/**
+ * A recovery strategy that will trigger if the parser failure is similar to that of some sample unfinished stream
+ * definition. The match is decided by analyzing the top frames of the stack trace emitted by the parser when it
+ * encounters the ill formed input. Multiple fingerprints are supported, as the control flow in the parser code may be
+ * different depending on the form of the expression. For example, for the rule {@code stream = module (| module)* },
+ * the pseudo code for the parser may look like
+ *
+ * <pre>
+ * <code>
+ * stream() {
+ *   module();  (1)
+ *   while(moreInput()) {
+ *     swallowPipe();
+ *     module();  (2)
+ *   }
+ * }
+ * </code>
+ * </pre>
+ *
+ * In that setup, whether we're dealing with the first module, or a subsequent module, stack frames would be different
+ * (see (1) and (2)).
+ *
+ * @author Eric Bottard
+ */
+public abstract class StacktraceFingerprintingRecoveryStrategy<E extends Exception> implements
+		RecoveryStrategy<E> {
+
+	private final Set<List<StackTraceElement>> fingerprints = new LinkedHashSet<>();
+
+	private final Class<E> exceptionClass;
+
+	/**
+	 * Construct a new StacktraceFingerprintingRecoveryStrategy given the parser, and the expected exception
+	 * class to be thrown for sample fragments of a stream definition that is to be parsed.
+	 *
+	 * @param exceptionClass the expected exception that results from parsing the sample fragment stream definitions.
+	 *        Stack frames from the thrown exception are used to store the fingerprint of this exception thrown by the
+	 *        parser.
+	 * @param samples the sample fragments of stream definitions.
+	 */
+	public StacktraceFingerprintingRecoveryStrategy(Class<E> exceptionClass,
+			String... samples) {
+		Assert.notNull(exceptionClass, "exceptionClass should not be null");
+		Assert.notEmpty(samples, "samples should not be null or empty");
+		this.exceptionClass = exceptionClass;
+		initFingerprints(samples);
+	}
+
+	@SuppressWarnings("unchecked")
+	private void initFingerprints(String... samples) {
+		for (String sample : samples) {
+			try {
+				new StreamDefinition("__dummy", sample);
+			}
+			catch (RuntimeException exception) {
+				if (this.exceptionClass.isAssignableFrom(exception.getClass())) {
+					computeFingerprint((E) exception);
+				}
+				else {
+					throw exception;
+				}
+			}
+		}
+	}
+
+	/**
+	 * Extract the top frames (until the call to the {@link StreamDefinition} constructor appears) of the given exception.
+	 */
+	private void computeFingerprint(E exception) {
```

Comment:
It does makes sense that you'd expect an "add..." method to accept the object that it is adding as an argument. At the same time, when I see a method called "compute" that returns void, it isn't clear to me what it is doing with the result that comes from the computation. IMO the fact that invoking this method mutates an internal data structure is not something that should be hidden. Perhaps `addFingerprintForException`? I'm open to other options.